lib/Doctrine/DBAL/Types/Type.php

```diff
@@ -42,50 +42,44 @@
-        self::SIMPLE_ARRAY => SimpleArrayType::class,
-        self::JSON_ARRAY => JsonArrayType::class,
-        self::JSON => JsonType::class,
-        self::OBJECT => ObjectType::class,
-        self::BOOLEAN => BooleanType::class,
-        self::INTEGER => IntegerType::class,
-        self::SMALLINT => SmallIntType::class,
-        self::BIGINT => BigIntType::class,
-        self::STRING => StringType::class,
-        self::TEXT => TextType::class,
-        self::DATETIME => DateTimeType::class,
-        self::DATETIME_IMMUTABLE => DateTimeImmutableType::class,
-        self::DATETIMETZ => DateTimeTzType::class,
+    private const BUILTIN_TYPES_MAP = [
+        self::TARRAY               => ArrayType::class,
+        self::SIMPLE_ARRAY         => SimpleArrayType::class,
+        self::JSON_ARRAY           => JsonArrayType::class,
+        self::JSON                 => JsonType::class,
+        self::OBJECT               => ObjectType::class,
+        self::BOOLEAN              => BooleanType::class,
+        self::INTEGER              => IntegerType::class,
+        self::SMALLINT             => SmallIntType::class,
+        self::BIGINT               => BigIntType::class,
+        self::STRING               => StringType::class,
+        self::TEXT                 => TextType::class,
+        self::DATETIME             => DateTimeType::class,
+        self::DATETIME_IMMUTABLE   => DateTimeImmutableType::class,
+        self::DATETIMETZ           => DateTimeTzType::class,
         self::DATETIMETZ_IMMUTABLE => DateTimeTzImmutableType::class,
-        self::DATE => DateType::class,
-        self::DATE_IMMUTABLE => DateImmutableType::class,
-        self::TIME => TimeType::class,
-        self::TIME_IMMUTABLE => TimeImmutableType::class,
-        self::DECIMAL => DecimalType::class,
-        self::FLOAT => FloatType::class,
-        self::BINARY => BinaryType::class,
-        self::BLOB => BlobType::class,
-        self::GUID => GuidType::class,
-        self::DATEINTERVAL => DateIntervalType::class,
+        self::DATE                 => DateType::class,
+        self::DATE_IMMUTABLE       => DateImmutableType::class,
+        self::TIME                 => TimeType::class,
+        self::TIME_IMMUTABLE       => TimeImmutableType::class,
+        self::DECIMAL              => DecimalType::class,
+        self::FLOAT                => FloatType::class,
+        self::BINARY               => BinaryType::class,
+        self::BLOB                 => BlobType::class,
+        self::GUID                 => GuidType::class,
+        self::DATEINTERVAL         => DateIntervalType::class,
     ];
 
+    /** @var TypeRegistry|null */
+    private static $typeRegistry;
+
     /**
-     * Prevents instantiation and forces use of the factory method.
+     * @internal Do not instantiate directly - use the factory method instead.
      */
-    final private function __construct()
+    final public function __construct()
```

Comment:
Is this method kept only for the sake of being marked `@internal`? As far as I understand, eventually `Type` should be converted to an interface and not have a constructor at all. Shoudn't it?

Reply:
It's to force empty constructor, but since Types are no longer instantiated from within themselves, it must be public.

Reply:
So this is only needed because the registry instantiates types. It's fine as a temporary solution, but it shouldn't be needed if instead of identifying types by classes, we identify them by instances implementing `Type`.

Reply:
That's my thought, once we only rely on Type interface (not the abstract class), we can remove restriction on the constructor. Right now we can't since we instantiate the types without parameters.

Reply:
Agreed.